### PR TITLE
Replace vendor tarballs with extracted directories and update to latest versions

### PR DESCRIPTION
**Pull request by raimue**
This commit replaces the vendor tarballs for tcl, tcllib, and thread with git submodules, based on the git mirrors of the upstream fossil repositories. If run inside a git repository, the configure script will clone and checkout the required sources.

This also adds a submodule for tclconfig, which is distributed in the release tarball of thread, but not in their repository. TclX is still in a tarball, as the upsteam repository is CVS and I could not find any canonical Git mirror.

### CODE REVIEW

**Comment by ryandesign** [comment]
This will address https://trac.macports.org/ticket/50811

**Comment by seanfarley** [comment]
Have you guys ever used submodules in your daily work? I ask because of the headache they cause and that there's a reason they're called A Feature of Last Resort.

**Comment by raimue** [comment]
I think some of the problems of submodules do not apply to this usage, because we will never modify any of the files inside the submodule. We could continue with tarballs, but as compressed data, this might take a lot of disk space in the history over time. We could fetch the tarballs with git-lfs, but that is even more complicated and generates a bootstrapping problem. We could import everything as plain files, but is that necessary if we are not going to modify them?

I am open for other proposals.

**Comment by seanfarley** [comment]
> I think some of the problems of submodules do not apply to this usage

Submodules always have problems :-) One of the biggest downsides to submodules is that the url has to exist (or redirect) for all time. It makes bisection should a pain in the ass.

You're right about git-lfs. That is also A Feature of Last Resort. For most teams, I recommend just vendoring the files. It's not that big and shouldn't be a problem. It can always be changed later as well without breaking bisection.

Alternatively, you could try subtrees. Here are two articles about the differences and usage:

https://blogs.atlassian.com/2013/05/alternatives-to-git-submodule-git-subtree/

https://developer.atlassian.com/blog/2015/05/the-power-of-git-subtree/

**Comment by neverpanic** [comment]
I am using submodules daily at work, and while they're annoying, our use case is mostly but a single update. Our Tcl trees don't change very often.

On the other hand, we mostly don't care about the history of Tcl and Tcllib, so we might as well use subtrees with squashing, which is more or less equal to just putting the files there with some added tooling.

File: aclocal.m4 [hunk 1]
```diff
@@ -314,9 +280,85 @@ AC_DEFUN([MP_CONFIG_TARBALL], [
...
+				AC_MSG_ERROR([Don't know how to extract tarball $mp_tarball])
+				;;
+		esac
+		(cd "$mp_tarball_vendordir"; umask 0022; "$mp_tarball_extract_cmd" -d < "$ac_abs_confdir/$mp_tarball" | tar xf - || AC_MSG_ERROR([failed to extract $mp_tarball]))
+	fi
+	if ! test -d "$ac_dir"; then
+		AC_MSG_ERROR([tarball $mp_tarball did not extract to $ac_dir])
+	fi
+
+	MP_CONFIG_SUBDIR(["$ac_dir"])
```
Comment (neverpanic) [inline aclocal.m4]: Should you pass on the configure parameters here or am I misunderstanding how the macro expansion works?

File: aclocal.m4 [hunk 2]
```diff
@@ -314,9 +280,85 @@ AC_DEFUN([MP_CONFIG_TARBALL], [
...
+dnl
+dnl MP_INIT_SUBMODULE([dir-of-submodule])
+AC_DEFUN([MP_INIT_SUBMODULE], [
+
+	ac_dir="$1"
+
+	if test -e ".git"; then
+		AC_PATH_PROG(GIT_BIN, [git], [])
+		AC_MSG_NOTICE([=== updating git submodule in $ac_dir])
+		"$GIT_BIN" submodule update --init -- $ac_dir || AC_MSG_ERROR([failed to update git submodule in $ac_dir])
```
Comment (neverpanic) [inline aclocal.m4]: `$ac_dir` should be quoted here.

File: configure [hunk 3]
```diff
@@ -9647,1253 +9647,31 @@ tcl_version=8.5.18
...
-			esac
-			mp_arg="--cache-file=$mp_sub_cache_file"
-					case $mp_arg in
-		*\'*)
-			mp_arg=$($as_echo "$mp_arg" | sed "s/'/'\\\\\\\\''/g")
-		;;
-	esac
-	mp_arg="'$mp_arg'"
-			mp_fn_var_prepend mp_sub_configure_args "$mp_arg "
+MP_CONFIG_SUBMODULE(vendor/tcl, unix, --prefix=${macports_libexec_dir} --enable-threads)
```
Comment (neverpanic) [inline configure]: You forgot to define `MP_CONFIG_SUBMODULE` so it shows up unexpanded in the generated configure script.

File: aclocal.m4 [hunk 4]
```diff
@@ -105,55 +105,21 @@ AC_DEFUN([MP_TOOL_PATH], [dnl
...
-	dnl Warning: Don't use $GZIP and $BZIP2 here, both tools interpret these
-	dnl env variables as additional parameters.
-	AC_PATH_PROG(GZIP_BIN, [gzip], [])
-	AC_PATH_PROG(BZIP2_BIN, [bzip2], [])
+dnl MP_CONFIG_SUBDIR([path-to-directory], [configure-parameters])
+AC_DEFUN([MP_CONFIG_SUBDIR], [
 
-	mp_tarball="$1"
-	ac_dir=$2
+	ac_dir="$1"
```
Comment (neverpanic) [inline aclocal.m4]: I think we should remove the quotes here, this becomes `ac_dir=""$1""` in the generated configure file, which I don't think I did on purpose. Same for the copies of this.

File: aclocal.m4 (hunk 1 above)
Comment (raimue) [inline aclocal.m4]: That was just a mistake. Interestingly, it still ended up in the right directory in the vendor-destroot.

File: aclocal.m4 (hunk 4 above)
Comment (raimue) [inline aclocal.m4]: I think the quotes here are correct, but there should not be another set of quotes when passing it to MP_CONFIG_SUBDIR.

**Comment by jmroot** [comment]
I would prefer extracting Tclx, and TBH just having extracted copies of the others in our repo would likely do the job without introducing more complexity. If we really want git to keep track of the connection to upstream, I'd go with subtrees.

**Comment by ryandesign** [comment]
If we have extracted copies, doesn't that introduce the temptation for someone to commit a change to those extracted files?

**Comment by jmroot** [comment]
If there is an issue with them, why shouldn't the files be changed?

**Comment by raimue** [comment]
I don't think git subtrees will be the right tool here, as upstream development is originally using fossil. So with git subtrees, we would just use a complicated way to produce the same result as extracting upstream tarballs.

My next proposal and compromise would be to use uncompressed upstream tarballs. This has the advantage that updating the tarball will not increase our repository size that much, as git will be able to compress them and de-duplicate blocks from previous versions. All files in the upstream tarball will be kept pristine, so we can easily throw them away on `make distclean`. If we need to, it will be easier to keep track of local patches on top of the upstream tarball instead of directly applying them onto the extracted directory, which is complicated to replicate on updates.

**Comment by larryv** [comment]
@ryandesign
> If we have extracted copies, doesn't that introduce the temptation for someone to commit a change to those extracted files?

In theory yes, but very few of us touch this repository at all, so I don't think it would be an issue in practice.

**Comment by raimue** [comment]
Any more input on my [last proposal](#issuecomment-264932607)? As this is now blocking #11 we should come to a conclusion.

**Comment by mojca** [comment]
I don't think I have any say in this and I won't have to deal with any consequences, but extracted copies sound fine to me.

I don't have any experience with submodules or subtrees other than knowing that subtrees make fetching tarballs from github a bit of a pain. I don't think there's any use in keeping uncompressed tarballs in the git repository in any case and I agree that we are not interested in history of upstream repositories.

I believe that we don't plan to commit anything there, but it won't be the end of the world if someone accidentally does, except that changes will be overwritten next time we upgrade to the next version. I don't think this will be a problem in practice, or at least not a big one anyway.

Submodules make for a cleaner history and would be slightly easier to update than when extracted copies are used, but we won't be updating the repositories more than once in a long while anyway.

**Comment by jmroot** [comment]
Sounds good, let's do it.

**Comment by neverpanic** [comment]
I'd prefer extracted files over uncompressed tarballs, but I'm fine with whatever you implement if it works. Let's just get it done.

**Comment by jmroot** [comment]
Oh, I thought that's what was being discussed. But reading @raimue's comment again I see that he did say uncompressed tarballs, not extracted. I would also lean towards extracting them, then.

**Comment by mojca** [comment]
Rainer, I know that this is slightly off-topic, but since you are working on this part of the code anyway, what about adding the xz library as well? Or is the main problem its licence? https://trac.macports.org/ticket/52000

**Comment by jmroot** [comment]
@raimue Will you be able to work on this in the near future or would you like me to work on it?

**Comment by ryandesign** [comment]
I don't understand why we want to keep third-party code in our repository, rather than using submodules as originally requested. This seems like exactly the purpose for which the submodule feature exists.

**Comment by raimue** [comment]
Summarizing the arguments:
 * Contra compressed tarballs: Significantly increases repo size with every update
 * Pro submodules: No third-party code in our repo
 * Contra submodules: Needs additional setup steps after cloning
 * Contra submodules: Full history increases the fetch size and disk usage, but we do not need it
 * Contra submodules: Upstream is not using git, so we would have to rely on mirrors
 * Contra submodules: If URLs of submodules ever change, bisecting gets harder
 * Pro subtrees: No additional history, no additional setup after cloning
 * Contra subtrees: Upstream is not using git, essentially the same as extracting tarballs
 * Pro uncompressed tarballs: Smaller increase of repo size on every update
 * Contra uncompressed tarballs: Additional complexity during configure
 * Pro extracted tarballs: Stores only upstream source differences in our repo history
 * Pro extracted tarballs: No additional setup steps
 * Contra extracted tarballs: Could lead to accidental edits (but unlikely)

The majority of the comments was in favor of extracted tarballs, so I will go with this and update this pull request accordingly.

**Comment by raimue** [comment]
Here is the updated pull request using extracted tarballs. I kept the branch name in order not to lose this discussion. I also folded #11 into this pull request.

I added symlinks in the vendor directory to at least keep some notion which upstream version we are using. That makes update commits a bit messier, though.

Another thing I noticed while doing this:
 * Contra extracted tarballs: Need to check and validate .gitignore on every update

**Comment by mojca** [comment]
> * Contra extracted tarballs: Need to check and validate .gitignore on every update

Even if we would use out-of-source builds?

**Comment by neverpanic** [comment]
This works for me:
```
diff --git a/vendor/Makefile.in b/vendor/Makefile.in
index 122b87cd..4a99c3f0 100644
--- a/vendor/Makefile.in
+++ b/vendor/Makefile.in
@@ -85,7 +85,6 @@ test:
 clean: clean-tcl clean-tcllib clean-tclx clean-tclthread
        rm -rf ${DESTROOT}

-distclean:
+distclean: distclean-tcl distclean-tcllib distclean-tclx distclean-tclthread
        rm -f Makefile
-       rm -rf $(TCL_SUBDIR:/unix=) $(TCLTHREAD_SUBDIR) $(TCLX_SUBDIR) $(TCLLIB_SUBDIR)
        rm -rf vendor-destroot
```

**Pull request merged**